sw/device/silicon_creator/rom/sigverify_keys.h

```diff
@@ -60,6 +64,96 @@ enum {
   kSigverifyNumEntriesPerOtpWord = sizeof(uint32_t),
 };
 
+/**
+ * Common initial sequence of public keys stored in ROM.
+ *
+ * OpenTitan ROM contains RSA and SPX keys whose definitions share this common
+ * initial sequence. This common initial sequence allows us to perform key
+ * lookup and validity checks in a generic manner by casting
+ * `sigverify_rom_rsa_key_t` or `sigverify_rom_spx_key_t` to this type.
+ */
+typedef struct sigverify_rom_key_common_sequence {
+  /**
+   * Type of the key.
+   */
+  sigverify_key_type_t key_type;
+  /**
+   * ID of the key.
+   */
+  uint32_t key_id;
+} sigverify_rom_key_common_sequence_t;
```

Comment:
I think this might be better expressed as `sigverify_rom_key_header`.
The way you're using this structure, you're expecting that `key_id` overlaps `data[0]` in the RSA key and SPX keys.  Although the offsets of the members work out, I don't know what the C spec has to say about this as `key_id` is `uint32_t` and `data` is `uint32_t []`.

I think it might be clearer to declare your keys as:
```
struct rsa_key {
  sigverify_key_type_t key_type;
  uint32_t data[RSA_KEY_LEN];
  uint32_t n0_inv[LEN];
};

struct spx_key_t {
  sigverify_key_type_t key_type;
  uint32_t data[SPX_KEY_LEN];
};
```

And then use a flexible array member in the union:
```
union rom_key_t {
  struct {
    sigverify_key_type_t key_type;
    uint32_t data[];
  } header;
  struct rsa_key rsa;
  struct spx_key spx;
};
```

This means that `data[0]` is always the key_id and that the `data` member is always the same type (`uint32_t []`) regardless of whether it is being accessed via `header`, `rsa` or `spx`.

Reply:
Thanks @cfrantz. I like `rom_key_header` better than `...common_sequence`. I'll rename shortly.

IIUC your second suggestion is to flatten the key definitions. I'd like to keep them as-is since `key_type` field does not have a meaning outside of key lookup and I'd prefer not leaking it to signature verification.

Regarding your last suggestion, we need a type for the header so that we can cast keys to that type. Did you inline it in the `union` for illustration purposes? Using a flexible array member is an interesting idea but @luismarques i this materially different than using an `uint32_t` and adding assertions for the offsets?